### PR TITLE
support launching Lightning ddp with traditional command

### LINKED ISSUE #7003: PyTorch Lightning ignores traditional WORLD_SIZE/RANK specifications in environment and doesn't document replacement
## 🐛 Bug

Standard torch.distributed environment variables seem to be handled differently.

### To Reproduce

```
$ MASTER_ADDR=192.168.1.3 MASTER_PORT=1234 WORLD_SIZE=2 RANK=0 python3 boring_model.py
```

### Expected behavior

Should wait for second job to connect to the MASTER_PORT. Instead, just starts training. No combination of arguments and environment variables seems to change this behavior.

It appears that PL handles startup for distributed processing differently somehow and internally overrides environment variables. This works pretty well for ddp on single nodes and (presumably) automates deployment under Slurm.

But I can't figure out from the documentation or from the samples how to arrange for the traditional PyTorch behavior in which all jobs are started up manually. 

For comparison, this code behaves as expected:

```
$ cat > simple.py
import torch

print("init")
torch.distributed.init_process_group("gloo")
print("done", torch.distributed.get_rank(), torch.distributed.get_world_size())
$ MASTER_ADDR=192.168.1.3 MASTER_PORT=1234 WORLD_SIZE=2 RANK=0 python3 simple.py & sleep 3; MASTER_ADDR=192.168.1.3 MASTER_PORT=1234 WORLD_SIZE=2 RANK=1 python3 simple.py
```

That is, the two jobs rendezvous as expected and then exit.

### Suggested Behavior

- document how the traditional PyTorch behavior can be reproduced (i.e., no calculations of nodes/size/... in DDP)
- perhaps provide some kind of command line flag that restores the traditional behavior (e.g., "--accelerator ddp_plain")
- maybe something else

### Environment

 - PyTorch Version (e.g., 1.0): 1.8.1
 - OS (e.g., Linux): Ubuntu 20.04
 - How you installed PyTorch (`conda`, `pip`, source): pip3
 - Build command you used (if compiling from source):
 - Python version: 3.8
 - CUDA/cuDNN version: 11.1
 - GPU models and configuration: 1080
 - Any other relevant information:

### PR DESCRIPTION
**Pull request by awaelchli**
## What does this PR do?

Fixes #7003

The Lightning environment provides a convenient way to launch DDP multi-gpu experiments. It launches the required number of processes automatically under the hood as explained in the docs. However, there is currently only a hacky way for the user to prevent this if they wish to launch all processes manually through the command line or with utilities like `torch.distributed.launch`. This PR adds detection of `LOCAL_RANK` variable in the environment and determines accordingly if processes need to be launched or not.

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
I made sure I had fun coding 🙃

### CODE REVIEW

File: pl_examples/bug_report_model.py [hunk 1]
```diff
@@ -53,11 +53,9 @@ def run():
     model = BoringModel()
     trainer = Trainer(
         default_root_dir=os.getcwd(),
-        limit_train_batches=1,
```
Comment (justusschock) [inline pl_examples/bug_report_model.py:56]: should this really be changed in the bug report model?

Comment (awaelchli) [reply pl_examples/bug_report_model.py:56]: nope :) will revert in final version

File: pytorch_lightning/plugins/training_type/ddp.py [hunk 2]
```diff
@@ -128,7 +128,7 @@ def _is_single_process_single_device(self) -> bool:
 
     def setup_environment(self):
         # start the other scripts
-        if not self.cluster_environment.creates_children() and os.environ.get("PL_IN_DDP_SUBPROCESS", "0") != "1":
+        if not self.cluster_environment.creates_children():
```
Comment (tchaton) [inline pytorch_lightning/plugins/training_type/ddp.py:131]: Neat !

**Review by tchaton — COMMENTED** [review]
Nice :)

**Review by carmocca — APPROVED** [review]
LGTM. Some comments:

Can you merge master? To make sure none of the recent changes have any impact.
Should we add a special test using `torch.distributed.launch`?

**Comment by awaelchli** [comment]
@carmocca adding a test for `torch.distributed.launch/run` would not really cover anything in this PR, as this would anyway invoke the torchelastic environment. What this PR is about is the manual way to launch individual processes (say in different terminals or a script). 

I added such calls to the `special_tests.sh` file, please have a look.

It adds about 10s of special test time.

**Review by ananthsub — APPROVED** [review]

File: tests/special_tests.sh [hunk 3]
```diff
@@ -78,6 +78,12 @@ if [ $? -eq 0 ]; then
     report+="Ran\ttests/utilities/test_warnings.py\n"
 fi
 
+# test that a user can manually launch individual processes
+args="--trainer.gpus 2 --trainer.accelerator ddp --trainer.max_epochs=1 --data.batch_size=32 --trainer.limit_train_batches=2 --trainer.limit_val_batches=2"
```
Comment (carmocca) [inline tests/special_tests.sh:82]: For the purpose of this test, this should be enough

```suggestion
args="--trainer.gpus 2 --trainer.accelerator ddp --trainer.fast_dev_run=True
```

Comment (awaelchli) [reply tests/special_tests.sh]: I took this from the existing tests

File: tests/special_tests.sh [hunk 4]
```diff
@@ -78,6 +78,12 @@ if [ $? -eq 0 ]; then
     report+="Ran\ttests/utilities/test_warnings.py\n"
 fi
 
+# test that a user can manually launch individual processes
```
Comment (tchaton) [inline tests/special_tests.sh:81]: Neat !

**Review by tchaton — APPROVED** [review]
LGTM !

File: tests/plugins/environments/test_lightning_environment.py [hunk 5]
```diff
@@ -37,6 +39,24 @@ def test_attributes_from_environment_variables():
...
 
+@pytest.mark.parametrize(
+    "environ, creates_children", [
+        ({}, False),
+        ({
+            "LOCAL_RANK": "2"
+        }, True),
+        ({
+            "NODE_RANK": "1"
+        }, False),
```
Comment (Borda) [inline tests/plugins/environments/test_lightning_environment.py:50]:
```suggestion
        (dict(LOCAL_RANK="2"), True),
        (dict(NODE_RANK="1"), False),
```

**Review by Borda — APPROVED** [review]

**Comment by awaelchli** [comment]
Yesterday GPU tests passed flawlessly. Of course it fails today when I get reviews.
RIP
